decisions/DR004-Core-language-support.md

```diff
@@ -0,0 +1,476 @@
+clear and manageable ownership semantics would facilitate support in the
+broadest range of future programming environments.
+
+Ideally, any bindings would also allow idiomatic programming within
+the target language, rather than forcing specific design patterns from
+any internals of OpenAssetIO - such as synchronous programming, RAII,
+etc.
+
+### Performance
+
+A core requirement raised by numerous parties is that the "happy path"
+should be as performant as possible. Thus, it is required that any
+mechanism introduced to broaden compatibility doesn't compromise the
+(current) majority deployment scenarios.
+
+Of note, a `C++` host talking to a `C++` manager has been the most
+frequently requested high-performance pathway.
+
+### Development lifecycle
+
+The overhead of maintaining dynamically linked `C++` libraries within a
+plugin-based architecture is often fraught with constantly chasing ABI
+changes and/or conflicts between mutual dependencies.
+
+Foundry has prior art in using `C` function pointer suites as a compiler
+isolation technique that can help de-couple the plugin host and the
+specifics of any given plugin.
+
+In the constantly evolving landscape defining which dependency versions
+are in use within which versions of common DCC tools, avoiding the need
+to matrix-compile and juggle plugin search paths per-host would be
+massively beneficial.
+
+## Options considered
+
+### Option 1
+
+`C++` and `Python` support only.
+
+**Estimated Cost**: Small
+
+#### Pros
+
+-   Simplifies development.
+
+#### Cons
+
+-   Leaves use within languages other than these a potentially
+    challenging exercise left to the reader.
+-   Compiler isolation support would need to be handled as an add-on
+    through an additional layer such as a `C` suite, `protobuf` or
+    similar.
+
+### Option 2
+
+`C` core library, bound to `C++` and `Python` as appropriate.
+
+**Estimated Cost**: Large
+
+#### Pros
+
+-   Makes future bindings or adoption in `C`-based languages far
+    simpler as all core API functionality is inherently available
+    through idiomatic `C`.
+-   Inherently presents a stable ABI isolation layer between hosts and
+    plugins.
+
+#### Cons
+
+-   The data model of OpenAssetIO _requires_ `map` and `optional` types.
+    Implementing these in pure `C` either requires significant coding,
+    or would add additional third-party dependencies. Either of these
+    add significantly to code complexity, maintenance overhead and
+    project risk.
+-   C is notoriously unsafe with regards to type-safety and memory
+    management. Higher-level systems programming languages, such as C++
+    and Rust, add protective mechanisms (both compile-time and run-time)
+    with little or no performance overhead.
+-   C is less expressive than higher-level languages, missing features
+    such as inheritance and generic programming, which would likely
+    increase development time.
```

Comment:
An earlier point in this document is that Foundry has prior art in C function suites, this point seems contradictory. The advertised lack of generic programmings stems from a rejection by the C++ community of macros; perhaps a better wording would be:

C programming idioms around inheritance and generic programming use macros. The use of macros makes for awkward, or impossible interoperation with languages such as C++ where the use of such constructs interfere with clarity and best practices.

Reply:
> An earlier point in this document is that Foundry has prior art in C function suites, this point seems contradictory
Foundry's C function pointer suites (structs) invariably call through to C++ logic. We could clarify that point if its causing confusion.

Reply:
Added a little clarification to the section on prior-art [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/175d30a3194aad501b825bb0fd02750d02588de4), and broadened to include memory isolation [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/5c28736890dffce8644ef24127afcec60421130f).

Reply:
Updated the "cons" of a complex `C` implementation [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/6111631839f7432e8fade7434e28e778be8203ed).